src/python/pants/backend/go/build.py

```diff
@@ -76,164 +103,294 @@ class GoBinaryFieldSet(PackageFieldSet):
+    request: BuildGoPackageRequest,
+) -> BuiltGoPackage:
+    wrapped_target = await Get(WrappedTarget, Address, request.address)
+    target = wrapped_target.target
+
+    if is_first_party_package_target(target):
+        source_files, resolved_package = await MultiGet(
+            Get(
+                SourceFiles,
+                SourceFilesRequest((target[GoPackageSources],)),
+            ),
+            Get(ResolvedGoPackage, ResolveGoPackageRequest(address=target.address)),
+        )
+        source_files_digest = source_files.snapshot.digest
+        source_files_subpath = target.address.spec_path
+    elif is_third_party_package_target(target):
+        module_path = target[GoExternalModulePath].value
+        if not module_path:
+            raise ValueError(f"_go_ext_mod_package at address {request.address} has a blank `path`")
+
+        module_version = target[GoExternalModuleVersion].value
+        if not module_version:
+            raise ValueError(
+                f"_go_ext_mod_package at address {request.address} has a blank `version`"
+            )
+
+        module, resolved_package = await MultiGet(
+            Get(
+                DownloadedExternalModule,
+                DownloadExternalModuleRequest(
+                    path=module_path,
+                    version=module_version,
+                ),
+            ),
+            Get(ResolvedGoPackage, ResolveExternalGoPackageRequest(address=request.address)),
+        )
+
+        source_files_digest = module.digest
+        source_files_subpath = resolved_package.import_path[len(module_path) :]
+    else:
+        raise ValueError(f"Unknown how to build Go target at address {request.address}.")
+
+    dependencies = await Get(Addresses, DependenciesRequest(field=target[Dependencies]))
+    dependencies_targets = await Get(Targets, Addresses(dependencies))
+    buildable_dependencies_targets = [
+        dep_tgt
+        for dep_tgt in dependencies_targets
+        if is_first_party_package_target(dep_tgt) or is_third_party_package_target(dep_tgt)
+    ]
+    built_go_deps = await MultiGet(
+        Get(BuiltGoPackage, BuildGoPackageRequest(tgt.address))
+        for tgt in buildable_dependencies_targets
+    )
+
+    gathered_imports = await Get(
+        GatheredImports,
+        GatherImportsRequest(
+            packages=FrozenOrderedSet(built_go_deps),
+            include_stdlib=True,
+        ),
+    )
 
     input_digest = await Get(
         Digest,
-        MergeDigests(input_root_digests),
+        MergeDigests([gathered_imports.digest, source_files_digest]),
     )
 
-    import_path = request.field_sets.import_path.value or ""
-    if not import_path:
-        raise ValueError("expected import_path to be non-empty")
+    import_path = resolved_package.import_path
     if request.is_main:
         import_path = "main"
 
+    # If there are any .s assembly files to be built, then generate a "symabis" file for consumption by Go compiler.
```

Comment:
Consider extracting this out into a helper rule so that this main rule is easier to read.

Reply:
TODO